client/driver/docker.go

```diff
@@ -190,13 +196,30 @@ type DockerDriverConfig struct {
 	ForcePull        bool                `mapstructure:"force_pull"`         // Always force pull before running image, useful if your tags are mutable
 	MacAddress       string              `mapstructure:"mac_address"`        // Pin mac address to container
 	SecurityOpt      []string            `mapstructure:"security_opt"`       // Flags to pass directly to security-opt
+	Devices          []DockerDevice      `mapstructure:"devices"`            // To allow mounting USB or other serial control devices
 }
 
 // Validate validates a docker driver config
 func (c *DockerDriverConfig) Validate() error {
 	if c.ImageName == "" {
 		return fmt.Errorf("Docker Driver needs an image name")
 	}
+	if len(c.Devices) > 0 {
+		for _, dev := range c.Devices {
+			if dev.HostPath == "" {
+				return fmt.Errorf("host path must be set in configuration for devices")
+			}
+
+			if dev.CgroupPermissions != "" {
+				for _, c := range dev.CgroupPermissions {
+					ch := string(c)
+					if ch != "r" && ch != "w" && ch != "m" {
+						return fmt.Errorf("invalid cgroup permission string: %q", dev.CgroupPermissions)
```

Comment:
Is this correct: https://github.com/docker/cli/blob/master/docs/reference/commandline/create.md#dealing-with-dynamically-created-devices---device-cgroup-rule?

Reply:
I made this validation logic after reading https://docs.docker.com/engine/reference/commandline/run/#add-host-device-to-container-device. Also, note this sentence at the end of that documentation: _Note: --device cannot be safely used with ephemeral devices. Block devices that may be removed should not be added to untrusted containers with --device._ The documentation you linked above for `--device-cgroup-rule` for dynamic devices, which looks like something else. I could be missing something though. 

We can create another follow up ticket to add configuration for `--device-cgroup-rule` without needing `--device` if this is something people need.

Reply:
Ah thanks for clarifying!